### PR TITLE
EZP-29541: As a developer, I want a API to manage user preference

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29541](https://jira.ez.no/browse/EZP-29541)
| **Bug/Improvement**| no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/UserPreferenceService.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     * List user preferences.
+     *
+     * @param int $offset the start offset for paging
+     * @param int $limit the number of user preferences returned
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     *
+     * @return \eZ\Publish\API\Repository\Values\UserPreference\UserPreferenceList
+     */
+    public function listPreferences(int $offset = 0, int $limit = 25): UserPreferenceList;
```
Comment (alongosz) [inline eZ/Publish/API/Repository/UserPreferenceService.php]: to conform to similar naming in ContentService - `loadPreferencesList`

File: eZ/Publish/API/Repository/Values/UserPreference/UserPreference.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * @property-read mixed $userId the user id this user preference belongs to
+ * @property-read string $name name of user preference
+ * @property-read string $value value of user preference
+ */
+abstract class UserPreference extends ValueObject
+{
+    /**
+     * ID of the user preference.
+     *
+     * @var mixed
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Values/UserPreference/UserPreference.php]: I really don't like ambiguity. Type here is well known.

File: eZ/Publish/API/Repository/Values/UserPreference/UserPreferenceSetStruct.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Values\UserPreference;
+
+use eZ\Publish\API\Repository\Values\ValueObject;
+
+class UserPreferenceSetStruct extends ValueObject
+{
+    /** @var int $userId */
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Values/UserPreference/UserPreferenceSetStruct.php]: CS: multiline PhpDoc.

File: eZ/Publish/API/Repository/UserPreferenceService.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+     * Get currently logged user preferences.
+     *
+     * @param int $offset the start offset for paging
+     * @param int $limit the number of user preferences returned
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     *
+     * @return \eZ\Publish\API\Repository\Values\UserPreference\UserPreferenceList
+     */
+    public function loadUserPreferences(int $offset = 0, int $limit = 25): UserPreferenceList;
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/API/Repository/UserPreferenceService.php]: @adamwojs I just talked to @webhdx. They also need listing. 

While right now our list is small, it's hard to determine 1) size of such list in Legacy 2) size of such list in the future. So due to scalability I'd keep pagination.

**Comment by mikadamczyk** [comment]
@adamwojs  @alongosz this one is ready for review

File: eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/Gateway/DoctrineDatabaseTest.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            ],
+        ], $results);
+    }
+
+    /**
+     * Return a ready to test DoctrineStorage gateway.
+     *
+     * @return \eZ\Publish\Core\Persistence\Legacy\UserPreference\Gateway\DoctrineDatabase
+     */
+    protected function getGateway(): DoctrineDatabase
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/Gateway/DoctrineDatabaseTest.php]: Please type-hint and refer to everywhere by abstract Gateway, not specific implementation.

File: eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/Gateway/DoctrineDatabaseTest.php [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+     * @param int $id
+     *
+     * @return array
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function loadUserPreference(int $id): array
+    {
+        $data = $this->connection
+            ->executeQuery('SELECT * FROM ezpreferences WHERE id = :id', ['id' => $id])
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/Gateway/DoctrineDatabaseTest.php]: Please use Doctrine Query Builder.
```php
$queryBuilder = $this->connection->createQueryBuilder();
$queryBuilder
    ->select('id', 'user_id', 'name', 'value')
    ->from('ezpreferences', 'p')
    ->where(
        $queryBuilder->expr()->eq(
            'p.id', 
            $queryBuilder->createPositionalParameter($id, ParameterType::INTEGER)
      )
;

return $queryBuilder->execute()->fetchAll(FechMode:::ASSOCIATIVE);
```
^ written in comment textarea, so can contain mistakes ;)

File: eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/HandlerTest.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+use eZ\Publish\Core\Persistence\Legacy\UserPreference\Handler;
+use eZ\Publish\SPI\Persistence\UserPreference\UserPreferenceSetStruct;
+use eZ\Publish\SPI\Persistence\UserPreference\UserPreference;
+use PHPUnit\Framework\TestCase;
+
+class HandlerTest extends TestCase
+{
+    const USER_PREFERENCE_ID = 1;
+
+    /** @var \eZ\Publish\Core\Persistence\Legacy\UserPreference\Gateway|\PHPUnit\Framework\MockObject\MockObject */
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Tests/UserPreference/HandlerTest.php]: Please use multiline PhpDoc (unless modifying some old class which already has this inconsistency)

File: eZ/Publish/Core/Persistence/Legacy/Tests/_fixtures/schema.sqlite.sql [hunk 8]
```diff
@@ -563,3 +563,13 @@ CREATE TABLE eznotification (
...
+
+CREATE TABLE ezpreferences (
+  id integer NOT NULL PRIMARY KEY AUTOINCREMENT,
+  name text(100) DEFAULT NULL,
+  user_id integer NOT NULL DEFAULT 0,
+  value text(255) DEFAULT NULL
+);
+
+CREATE INDEX ezpreferences_name ON ezpreferences (name);
+CREATE INDEX ezpreferences_user_id_idx ON ezpreferences (user_id,name);
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Tests/_fixtures/schema.sqlite.sql]: file consistency / code style: space after comma

File: eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+        if (0 < count($userPreferences)) {
+            $currentUserPreference = reset($userPreferences);
+            $currentUserPreferenceId = (int)$currentUserPreference['id'];
+
+            $query
+                ->update(self::TABLE_USER_PREFERENCES)
+                ->set(self::COLUMN_VALUE, ':value')
+                ->where($query->expr()->eq(self::COLUMN_ID, ':id'))
+                ->setParameter(':id', $currentUserPreferenceId, PDO::PARAM_INT)
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php]: `, ParameterType::INTEGER`

File: eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                ->setParameter(':value', $userPreference->value, PDO::PARAM_STR);
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php]: ` ParameterType::STRING`
(AFAIR)

File: eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            ->from(self::TABLE_USER_PREFERENCES)
+            ->where($query->expr()->eq(self::COLUMN_USER_ID, ':user_id'))
+            ->setFirstResult($offset);
+
+        if ($limit > 0) {
+            $query->setMaxResults($limit);
+        }
+
+        $query->orderBy(self::COLUMN_ID, 'ASC');
+        $query->setParameter(':user_id', $userId, PDO::PARAM_INT);
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php]: `ParameterType::`

File: eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+     * {@inheritdoc}
+     */
+    public function countUserPreferences(int $userId): int
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->select('COUNT(' . self::COLUMN_ID . ')')
+            ->from(self::TABLE_USER_PREFERENCES)
+            ->where($query->expr()->eq(self::COLUMN_USER_ID, ':user_id'))
+            ->setParameter(':user_id', $userId, PDO::PARAM_INT);
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php]: `ParameterType::`

File: eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function countUserPreferences(int $userId): int
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->select('COUNT(' . self::COLUMN_ID . ')')
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/UserPreference/Gateway/DoctrineDatabase.php]:
```php
->select(
    $this->connection->getDatabasePlatform()->getCountExpression(self::COLUMN_ID);
)
```

File: eZ/Publish/Core/Repository/Tests/Service/Mock/UserPreferenceTest.php [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+                    return new UserPreference();
+                });
+        });
+
+        $this->createAPIUserPreferenceService()->setUserPreference([$apiUserPreferenceSetStruct]);
+    }
+
+    /**
+     * @covers \eZ\Publish\Core\Repository\UserPreferenceService::setUserPreference
+     * @expectedException \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/UserPreferenceTest.php]: Hmm, I think you're the one who showed us PHPUnit's `$this->expectException()`, why go back to old standard? ;)

File: eZ/Publish/Core/settings/user_preference.yml [hunk 15]
```diff
@@ -0,0 +1,3 @@
+services:
+    eZ\Publish\API\Repository\UserPreferenceService:
+        alias: eZ\Publish\Core\SignalSlot\UserPreferenceService
```
Comment (alongosz) [inline eZ/Publish/Core/settings/user_preference.yml]: Missing LF at EOF

File: eZ/Publish/SPI/Persistence/UserPreference/UserPreference.php [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+     * ID of the user preference.
+     *
+     * @var int
+     */
+    public $id;
+
+    /**
+     * The ID of the user this user preference belongs to.
+     *
+     * @var mixed
```
Comment (alongosz) [inline eZ/Publish/SPI/Persistence/UserPreference/UserPreference.php]: I still believe that `$userId` will be always int, especially due to enforcing of strict types in the implementation :P

File: eZ/Publish/Core/Repository/Tests/Service/Mock/UserPreferenceTest.php (hunk 14 above)
Comment (mikadamczyk) [reply eZ/Publish/Core/Repository/Tests/Service/Mock/UserPreferenceTest.php]: Because of respect to consistency with other tests :)

Comment (alongosz) [reply eZ/Publish/Core/Repository/Tests/Service/Mock/UserPreferenceTest.php]: Consistency is tricky. We started some time ago to introduce ->expectException to kernel tests as well, but you have a good point. 
Let's keep it here as is and decide later on how to approach such improvements.

**Comment by lserwatka** [comment]
QA will done as part of user preferences

**Pull request merged**